File: apps/desktop/electron/window.ts
/**
 * 窗口编排与 app:// 静态托管。
 *
 * 静态导出经自定义 `app://sag/` 协议托管（file:// 无法承载绝对路径资源，
 * 且需要稳定 origin 供后端 CORS 白名单，ADR-0023）。
 * 运行时配置以 additionalArguments → preload contextBridge 注入，
 * 先于一切页面脚本（ADR-0007）；本地访问密钥永不进渲染进程（ADR-0011）。
 */

import { createReadStream, existsSync, statSync } from "node:fs";
import { join, normalize, sep } from "node:path";
import { pathToFileURL } from "node:url";
import { BrowserWindow, app, net, protocol } from "electron";

export const APP_SCHEME = "app";
export const APP_HOST = "sag";
export const APP_ORIGIN = `${APP_SCHEME}://${APP_HOST}`;

let mainWindow: BrowserWindow | null = null;
let bootWindow: BrowserWindow | null = null;

/** 必须在 app ready 之前调用。 */
export function registerAppScheme(): void {
  protocol.registerSchemesAsPrivileged([
    {
      scheme: APP_SCHEME,
      privileges: {
        standard: true,
        secure: true,
        supportFetchAPI: true,
        corsEnabled: true,
        stream: true,
      },
    },
  ]);
}

function webRoot(): string {
  return app.isPackaged
    ? join(process.resourcesPath, "web")
    : join(__dirname, "..", "dist-web");
}

/** app ready 之后调用：app://sag/<path> → dist-web 文件（目录索引 + 404 回退）。 */
export function serveAppProtocol(): void {
  const root = webRoot();
  protocol.handle(APP_SCHEME, (request) => {
    const url = new URL(request.url);
    let pathname = decodeURIComponent(url.pathname);
    if (pathname.endsWith("/")) pathname += "index.html";
    if (pathname === "") pathname = "/index.html";
    // 路径遏制：解析后必须仍在 web 根内
    const resolved = normalize(join(root, pathname));
    if (resolved !== root && !resolved.startsWith(root + sep)) {
      return new Response("forbidden", { status: 403 });
    }
    const candidates = [resolved, `${resolved}.html`, join(resolved, "index.html")];
    for (const candidate of candidates) {
      if (existsSync(candidate) && statSync(candidate).isFile()) {
        return net.fetch(pathToFileURL(candidate).toString());
      }
    }
    const notFound = join(root, "404.html");
    if (existsSync(notFound)) {
      return net.fetch(pathToFileURL(notFound).toString()).then(
        (response) => new Response(response.body, { status: 404, headers: response.headers }),
      );
    }
    return new Response("not found", { status: 404 });
  });
}

function preloadPath(): string {
  return join(__dirname, "preload.js");
}

export function createBootWindow(): BrowserWindow {
  if (bootWindow && !bootWindow.isDestroyed()) return bootWindow;
  bootWindow = new BrowserWindow({
    width: 460,
    height: 420,
    resizable: false,
    maximizable: false,
    title: "SAG",
    show: true,
    webPreferences: {
      preload: preloadPath(),
      contextIsolation: true,
      nodeIntegration: false,
      sandbox: true,
    },
  });
  bootWindow.on("closed", () => {
    bootWindow = null;
  });
  void bootWindow.loadURL(`${APP_ORIGIN}/boot.html`);
  return bootWindow;
}

/** sidecar 就绪：以最终端口创建工作台窗口（配置经 additionalArguments 注入）并收起 boot。 */
export function showMainWindow(port: number): BrowserWindow {
  if (mainWindow && !mainWindow.isDestroyed()) {
    mainWindow.show();
    mainWindow.focus();
  } else {
    const runtimeConfig = {
      apiBase: `http://127.0.0.1:${port}`,
      host: "desktop",
      appVersion: app.getVersion(),
      flags: { enableWindowScaling: false },
    };
    mainWindow = new BrowserWindow({
      width: 1280,
      height: 800,
      minWidth: 960,
      minHeight: 640,
      title: "SAG",
      show: false,
      webPreferences: {
        preload: preloadPath(),
        contextIsolation: true,
        nodeIntegration: false,
        sandbox: true,
        additionalArguments: [`--sag-runtime-config=${JSON.stringify(runtimeConfig)}`],
      },
    });
    mainWindow.once("ready-to-show", () => mainWindow?.show());
    mainWindow.on("closed", () => {
      mainWindow = null;
    });
    void mainWindow.loadURL(`${APP_ORIGIN}/index.html`);
  }
  if (bootWindow && !bootWindow.isDestroyed()) bootWindow.close();
  return mainWindow;
}

/** 终态错误 / 崩溃循环时回到 boot 恢复页。 */
export function showBootWindow(): void {
  if (mainWindow && !mainWindow.isDestroyed()) {
    mainWindow.destroy();
    mainWindow = null;
  }
  const window = createBootWindow();
  window.show();
  window.focus();
}

/** 显示并聚焦任一现存窗口（托盘「打开」/二次启动/Dock 激活共用）。 */
export function focusAnyWindow(): void {
  const target =
    (mainWindow && !mainWindow.isDestroyed() && mainWindow) ||
    (bootWindow && !bootWindow.isDestroyed() && bootWindow) ||
    null;
  if (target) {
    target.show();
    if (target.isMinimized()) target.restore();
    target.focus();
    return;
  }
  showBootWindow(); // 两个窗口都不在的边界:回到 boot 页兜底
}

export function getMainWindow(): BrowserWindow | null {
  return mainWindow && !mainWindow.isDestroyed() ? mainWindow : null;
}

export function broadcastToWindows(channel: string, payload: unknown): void {
  for (const window of BrowserWindow.getAllWindows()) {
    if (!window.isDestroyed()) window.webContents.send(channel, payload);
  }
}

File: apps/desktop/src-tauri/src/window.rs
//! 窗口编排：boot（启动/恢复页）与 main（工作台）。
//!
//! 运行时配置必须在窗口创建时以 initialization_script 注入（ADR-0007），
//! 端口变化（用户改端口重试）需要重建窗口 —— 因此 boot 与 main 分离：
//! boot 不带配置脚本，main 在 sidecar 就绪、端口定格后才创建。

use tauri::{AppHandle, Manager, WebviewUrl, WebviewWindowBuilder};

use crate::runtime_config;

pub const BOOT_LABEL: &str = "boot";
pub const MAIN_LABEL: &str = "main";

pub fn create_boot_window(app: &AppHandle) -> tauri::Result<()> {
    if app.get_webview_window(BOOT_LABEL).is_some() {
        return Ok(());
    }
    WebviewWindowBuilder::new(app, BOOT_LABEL, WebviewUrl::App("boot.html".into()))
        .title("SAG")
        .inner_size(460.0, 400.0)
        .resizable(false)
        .maximizable(false)
        .center()
        .build()?;
    Ok(())
}

/// sidecar 就绪后调用：创建工作台窗口（带运行时配置注入）并收起 boot 窗口。
pub fn show_main_window(app: &AppHandle, port: u16) -> tauri::Result<()> {
    if let Some(existing) = app.get_webview_window(MAIN_LABEL) {
        let _ = existing.show();
        let _ = existing.set_focus();
    } else {
        WebviewWindowBuilder::new(app, MAIN_LABEL, WebviewUrl::App("index.html".into()))
            .title("SAG")
            .inner_size(1280.0, 800.0)
            .min_inner_size(960.0, 640.0)
            .initialization_script(&runtime_config::build_init_script(port))
            .build()?;
    }
    if let Some(boot) = app.get_webview_window(BOOT_LABEL) {
        let _ = boot.close();
    }
    Ok(())
}

/// 终态错误 / 崩溃循环时回到 boot 恢复页。
pub fn show_boot_window(app: &AppHandle) {
    if let Some(main) = app.get_webview_window(MAIN_LABEL) {
        let _ = main.close();
    }
    if let Err(error) = create_boot_window(app) {
        tracing::error!(%error, "创建启动窗口失败");
        return;
    }
    if let Some(boot) = app.get_webview_window(BOOT_LABEL) {
        let _ = boot.show();
        let _ = boot.set_focus();
    }
}

/// 显示并聚焦任一现存窗口（托盘「打开」与二次启动聚焦共用）。
pub fn focus_any_window(app: &AppHandle) {
    for label in [MAIN_LABEL, BOOT_LABEL] {
        if let Some(window) = app.get_webview_window(label) {
            let _ = window.show();
            let _ = window.unminimize();
            let _ = window.set_focus();
            return;
        }
    }
    // 两个窗口都不在（main 曾被隐藏后关闭等边界）：回到 boot 页兜底
    show_boot_window(app);
}
